Add tutorials using normalizing flows

francois-rozet opened the pull request:
This PR adds tutorials (SVI and VAE) using normalizing flows. The normalizing flow implementations are loaded from the [Zuko](https://github.com/probabilists/zuko) library. By the way, I think it could be interesting to add a "compatibility with external tools" section to the examples.

In addition (as discussed in #3297), this PR also updates the normalizing flow tutorial to mention that Pyro's built-in flow API is not developed anymore. I also fixed a mistake ($\det | \frac{\partial f}{\partial x} |$ should be $| \det \frac{\partial f}{\partial x} |$) in the math and cleaned up a bit.

martinjankowiak commented:
@francois-rozet isn't `rsample_and_log_prob` zuko-specific?

francois-rozet commented:
Indeed, but the class still works if `dist` does not implement `rsample_and_log_prob`. So it is not actually necessary to import `zuko`. Do you have a preference for `pyro.distributions` or `contrib.zuko` ?

BTW, `Transform.call_and_ladj` (also Zuko-specific) and `Distribution.rsample_and_log_prob` could be worth moving upstream (to torch itself), as they allow sampling and scoring without cache.

francois-rozet commented on tutorial/source/normalizing_flows_intro.ipynb:
```diff
@@ -4,10 +4,12 @@
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
+    "\n",
+    "> The development of Pyro's built-in flows has stopped in favor of external libraries. We recommend [Zuko](https://zuko.readthedocs.io) as it is compatible with Pyro, implements many flow architectures and is [well documented](https://zuko.readthedocs.io).\n",
```
I agree. I wanted to do that initially, but some libraries are unmaintained (`nflows`, `FlowTorch`) and some I don't know how they integrate with Pyro (`FrEIA`, `normflows`) as they don't follow PyTorch's distribution API. So should I just put all of them? I can also mention the [awesome-normalizing-flows](https://github.com/janosh/awesome-normalizing-flows) repo.

For the NF tuto, I only made minor edits, which either fix the error mentioned previously or simplify math notations (removing parentheses when clear from context). For example,

![image](https://github.com/pyro-ppl/pyro/assets/37352336/202610a7-d8e4-453e-ab79-e360ba80da59)


became

![image](https://github.com/pyro-ppl/pyro/assets/37352336/047a7323-ec68-4711-b145-3acbcea132a6)

or

![image](https://github.com/pyro-ppl/pyro/assets/37352336/b5f060cb-170b-4f9f-8b6b-3a6dab89a775)


became

![image](https://github.com/pyro-ppl/pyro/assets/37352336/9ead8716-d6c4-4564-95c9-c87ef29304d7)

My editor also removed trailing white-spaces, but nothing else changed.

martinjankowiak replied:
i defer to your judgment as to which libraries are worth mentioning, though i'd err on the side of more. you can just make an explicit caveat like "although we note that some of these libraries may no longer be actively maintained and may have APIs that are not directly compatible with Pyro (and thus may require a custom wrapper for use in Pyro)."

martinjankowiak commented:
@fritzo do you have a preference between `pyro.distributions` and `contrib.zuko`? i do not. the former is probably more discoverable...

martinjankowiak commented:
note you can put a test for your contrib code in `tests/contrib`

martinjankowiak commented on pyro/contrib/zuko.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class Zuko2Pyro(pyro.distributions.TorchDistribution):
+    r"""Wraps a Zuko (or PyTorch) distribution as a Pyro distribution."""
```
can you expand this docstring? link to zuko? etc

also the `(or PyTorch)` seems confusing since we have `pyro.distributions.TorchDistribution` for that

francois-rozet replied:
I don't think `TorchDistribution` does the same think. `Zuko2Pyro` takes any **instance** of distribution and makes it a valid Pyro distribution.

I'll add a page in the docs.

martinjankowiak commented on pyro/contrib/zuko.py:
```diff
@@ -0,0 +1,46 @@
+
+    def log_prob(self, x: Tensor) -> Tensor:
+        if x in self.cache:
+            return self.cache[x]
```
@fritzo can you please take a look at the caching logic?

francois-rozet replied:
Note that I don't clear the cache as I consider that `Zuko2Pyro` objects should be temporary.

francois-rozet replied:
Note that the `cache` is only used when `x` is sampled with `rsample_and_log_prob`. But, yes I will mention it in the docstring.

martinjankowiak commented on tests/contrib/test_zuko.py:
```diff
@@ -0,0 +1,56 @@
+
+
+@pytest.mark.parametrize("multivariate", [True, False])
+def test_Zuko2Pyro(multivariate: bool):
```
@fritzo anything else worth testing for?

francois-rozet replied:
I can add a "fake" `rsample_and_log_prob` method to the `dist` object and check that it runs with that.

```python
def rsample_and_log_prob(self, shape):
    x = self.rsample(shape)
    return x, self.log_prob(x)

dist.rsample_and_log_prob = rsample_and_log_prob
```

This would at least test that `ZukoToPyro` works as expected.

martinjankowiak replied:
sounds like a reasonable solution to me

francois-rozet replied:
It is done!

martinjankowiak commented on tutorial/source/normalizing_flows_intro.ipynb:
```diff
@@ -4,10 +4,12 @@
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
+    "\n",
+    "> The development of Pyro's built-in flows has stopped in favor of external libraries, such as [Zuko](https://github.com/probabilists/zuko), [nflows](https://github.com/bayesiains/nflows), [normflows](https://github.com/VincentStimper/normalizing-flows) or [FlowTorch](https://github.com/stefanwebb/flowtorch). Some of these libraries may have interfaces that are not directly compatible with P ...
```
think this should point to flowtorch.ai?

martinjankowiak commented on tutorial/source/normalizing_flows_intro.ipynb:
```diff
@@ -4,10 +4,12 @@
-    "# Normalizing Flows - Introduction (Part 1)\n",
+    "# Normalizing Flows - Introduction\n",
+    "\n",
+    "This tutorial introduces Pyro's built-in normalizing flows. It is independent of most of Pyro, but users may want to read about distribution shapes in the [Tensor Shapes Tutorial](http://pyro.ai/examples/tensor_shapes.html).\n",
```
can you add something like this to the top of both example notebooks?

"In this notebook we leverage normalizing flows implemented in Zuko, but similar functionality could also be obtained with another PyTorch-based normalizing library."

martinjankowiak commented on tutorial/source/svi_flow_guide.ipynb:
```diff
@@ -0,0 +1,236 @@
+   "source": [
+    "## SVI\n",
+    "\n",
+    "We train our guide with a standard stochastic variational inference (SVI) pipeline. We use 16 particles to reduce the variance of the ELBO and clip the norm of the gradients to make training more stable."
```
quite aggressive clipping?

francois-rozet replied:
That's a value I use quite often, but I don't have much intuition about it.

martinjankowiak commented on tutorial/source/svi_flow_guide.ipynb:
```diff
@@ -0,0 +1,236 @@
+    "        z = pyro.sample(\"z\", prior)\n",
+    "\n",
+    "        with pyro.plate(\"obs\", 5):\n",
+    "            pyro.sample(\"x\", likelihood(z), obs=x.transpose(0, 1))"
```
are all these transposes necessary?

francois-rozet replied:
Either there or in the guide.

martinjankowiak replied:
it seems like this could be re-organized more elegantly to reduce the number of transposes

francois-rozet replied:
I think I can reduce the number of `transpose` from 2 to 1.

martinjankowiak commented:
@fritzo do we need to add zuko to `EXTRAS_REQUIRE`?

francois-rozet commented:
@martinjankowiak I have addressed your comments (at least the ones I could). I don't think it is necessary to add zuko in `EXTRAS_REQUIRE` as even `contrib/zuko.py` does not rely on it. However, it seems like `doctest` tries to run the example I added in the docstring and fails because it does not import `zuko`. I can modify the example to be a `code-block`, which should not be executed.

martinjankowiak reviewed (dismissed):
lgtm. @fritzo could you please take a final pass, especially w.r.t. caching logic?

fritzo reviewed (commented):
@francois-rozet thank you for cleaning up our existing tutorials and for adding this new tutorial! I think these will help lots of users provide exposure for Zuko. And thanks for your patience in this very slow holiday review cycle. I think we can push a Pyro 1.9 release soon after this merges.

LGTM, subject to @martinjankowiak's review and a couple nits.

francois-rozet commented on pyro/contrib/zuko.py:
```diff
@@ -0,0 +1,77 @@
+import pyro
+
+
+class Zuko2Pyro(pyro.distributions.TorchDistribution):
```
Ahah I thought it was "cool" :laughing:, will fix

francois-rozet commented:
I addressed your comments, and I think the tests have passed (went to bed before the end).

martinjankowiak commented:
lgtm apart from possibly adding a `rsample_and_log_prob` test

francois-rozet commented:
> apart from possibly adding a rsample_and_log_prob test

I added it in the test.

martinjankowiak commented:
sorry i missed that thanks

francois-rozet commented:
🎉

martinjankowiak commented:
thanks @francois-rozet !!
